src/System.Memory/src/System/Memory.cs

```diff
@@ -193,6 +193,7 @@ public unsafe MemoryHandle Retain(bool pin = false)
                 if (_index < 0)
                 {
                     memoryHandle = ((OwnedMemory<T>)_arrayOrOwnedMemory).Pin();
+                    memoryHandle.AddOffset((_index & RemoveOwnedFlagBitMask) * Unsafe.SizeOf<T>());
```

Comment:
It feels wrong to me to expose a public method like AddOffset for mutating the handle.  I think I'd prefer either:
- Just have AddOffset be internal, or replace it with an internal SetPointer and use Unsafe.Add to compute the new one
- Create a new MemoryHandle with the inputs from the one returned by Pin, and just don't dispose the first one.

Reply:
I agree, creating new memory handle sound like much better idea and does not require any API change.